### PR TITLE
#12371: Migrate `moreh_getitem` operation from `tt_eager` to `ttnn` 

### PR DESCRIPTION
**Pull request by thd1007**
### Ticket
#12371 

### Problem description
`moreh_getitem` was deprecated alongside `tt_dnn`. This PR ports it to `ttnn` using its operation format.

### What's changed
- Removed `moreh_getitem`'s tt_dnn implementations
- Move device code to `ttnn`
- Create new wrapper code for `ttnn` with new modules

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_getitem/moreh_getitem_pybind.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    bind_registered_operation(
+        module,
+        ttnn::moreh_getitem,
+        "Moreh moreh_getitem operation",
+        ttnn::pybind_arguments_t{
+            py::arg("input"),
+            py::arg("index_tensors"),
+            py::arg("index_dims"),
+            py::arg("output") = std::nullopt,
+            py::arg("output_mem_config") = std::nullopt});
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_getitem/moreh_getitem_pybind.cpp:21]: >In ttnn we use kw_only arguments dtype and memory_config.
please adjust for consistency.
Also make sure that arguments which are kw args in torch are also kw args here.
Thank you!


https://github.com/tenstorrent/tt-metal/pull/12294#discussion_r1750369085

File: ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_tilized_factory.cpp [hunk 2]
```diff
@@ -82,49 +82,46 @@ operation::ProgramWithCallbacks moreh_getitem_tilized(
...
 
         Program program = Program();
 
         // create circular buffers
-        auto src_cb_data_format = tt_metal::datatype_to_dataformat_converter(input.get_dtype());
-        auto index_cb_data_format = tt_metal::datatype_to_dataformat_converter(index_tensors.at(0).get_dtype());
-        auto output_cb_data_format = tt_metal::datatype_to_dataformat_converter(output.get_dtype());
+        auto src_cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(input.get_dtype());
+        auto index_cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(index_tensors[0].get_dtype());
+        auto output_cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(output.get_dtype());
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_tilized_factory.cpp:117]:
```suggestion
        using namespace tt;
        using namespace tt::tt_metal;
    
        auto src_cb_data_format = datatype_to_dataformat_converter(input.get_dtype());
        auto index_cb_data_format = datatype_to_dataformat_converter(index_tensors[0].get_dtype());
        auto output_cb_data_format = datatype_to_dataformat_converter(output.get_dtype());
```
https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/examples/example/device/single_core_program_factory.cpp#L13

adding `using namespace tt;` could reduce the amount of code.

File: ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.hpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#include "ttnn/decorators.hpp"
+#include "ttnn/device_operation.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::moreh::moreh_getitem {
+struct MorehGetItemOperation {
+    struct operation_attributes_t {
+        const std::vector<uint32_t> index_dims;
+        // const CoreRange core_range;
+        const MemoryConfig output_mem_config;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.hpp:19]:
```suggestion
        const MemoryConfig output_memory_config;
```
It seems that TT prefers `memory_config` over `mem_config`

Comment (hschoi4448) [reply ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.hpp:19]: I will approve once this is fixed.

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**